vsp.go

```diff
@@ -57,65 +61,315 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+			}
+
+			tip := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			w := vsp.w.Internal()
+
+			// Don't perform any actions while transactions are not synced through
+			// the tip block.
+			rp, err := w.RescanPoint(ctx)
+			if err != nil {
+				return err
+			}
+			if rp != nil {
+				log.Debugf("Skipping autobuyer actions: transactions are not synced on Wallet: %s", vsp.w.Name)
+				continue
+			}
+
+			tipHeader, err := w.BlockHeader(ctx, tip)
+			if err != nil {
+				log.Error(err)
+				continue
+			}
+			height := int32(tipHeader.Height)
+
+			// Cancel any ongoing ticket purchases which are buying
+			// at an old ticket price or are no longer able to
+			// create mined tickets the window.
+			if height+2 >= nextIntervalStart {
+				for i, cancel := range cancels {
+					cancel()
+					cancels[i] = nil
+				}
+				cancels = cancels[:0]
+
+				intervalSize := int32(w.ChainParams().StakeDiffWindowSize)
+				currentInterval := height / intervalSize
+				nextIntervalStart = (currentInterval + 1) * intervalSize
+
+				// Skip this purchase when no more tickets may be purchased in the interval and
+				// the next sdiff is unknown.  The earliest any ticket may be mined is two
+				// blocks from now, with the next block containing the split transaction
+				// that the ticket purchase spends.
+				if height+2 == nextIntervalStart {
+					log.Debugf("Skipping purchase: next sdiff interval starts soon")
+					continue
+				}
+				// Set expiry to prevent tickets from being mined in the next
+				// sdiff interval.  When the next block begins the new interval,
+				// the ticket is being purchased for the next interval; therefore
+				// increment expiry by a full sdiff window size to prevent it
+				// being mined in the interval after the next.
+				expiry = nextIntervalStart
+				if height+1 == nextIntervalStart {
+					expiry += intervalSize
+				}
+			}
+
+			wal := vsp.w
+			// Get the account balance to determine how many tickets to buy
+			bal, err := wal.GetAccountBalance(int32(vsp.purchaseAccount))
+			if err != nil {
+				return err
+			}
+
+			spendable := bal.Spendable
+			if spendable < balanceToMaintain {
+				log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+				return nil
+			}
+
+			spendable -= balanceToMaintain
+			sdiff, err := wal.Internal().NextStakeDifficultyAfterHeader(ctx, tipHeader)
+			if err != nil {
+				return err
+			}
+
+			buy := int(dcrutil.Amount(spendable) / sdiff)
+			if buy == 0 {
+				log.Debugf("Skipping purchase: low available balance on Wallet: %s", wal.Name)
+				return nil
+			}
+
+			max := int(wal.Internal().ChainParams().MaxFreshStakePerBlock)
+			if buy > max {
+				buy = max
+			}
+
+			multiple := buy // multiple is the max amount of tickets that can be bought based on available funds
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			buyTickets := func() {
+				err := vsp.buyTickets(cancelCtx, expiry, sdiff, passphrase)
```

Comment:
dcrwallet uses this signature `cancelCtx, passphrase, tipHeader, expiry, &cfg`, shouldn't differ too much, this can be
```suggestion
				err := vsp.buyTickets(cancelCtx, passphrase, sdiff, expiry)
```

with `sdiff` replacing `tipHeader` for obvious reasons.